llama.cpp

```diff
@@ -2162,7 +2163,11 @@ struct llama_vocab {
     std::unordered_map<token, id> token_to_id;
     std::vector<token_data>       id_to_token;
 
-    std::vector<id> special_tokens_cache;
+    bool has_cache = false;
```

Comment:
Is there a mechanism by which the vocab can be loaded _without_ having a cache in place? If not, I'm wondering why `has_cache` is useful right now...?

Reply:
There was a way to exit early before creating the cache if the tokenizer was unknown. I've removed this path by throwing an exception: [1494a18](https://github.com/ggerganov/llama.cpp/pull/7587/commits/1494a1841ef999cf3a6c4aa5a30ffd2d689a8b79)

There is another path where the GGUF explicitly does not contain a vocabulary: `"no_vocab"`. In that case calling any of the functions that rely on a cache would throw exception due to accessing the caches via `cache.at()`. I think this makes sense

Removed `has_cache` and replaced the unordered maps with vectors

Reply:
Perfect, thank you!

I like all your changes here -- this all feels really good. The only other thing that I'll note is the caveat that I noted on @ochafik 's similar PR in #6811 : 

https://github.com/ggerganov/llama.cpp/pull/6811#discussion_r1589380140

> > I think it'd be simpler to leave it as is and keep it as an area where to potentially squeeze a couple of MB when times are scarce. wdyt?

> This also sounds not unreasonable, but I don't know how to weigh such things. I know I really like grammar-constrained sampling, but I don't know how popular the feature is overall, and is it worth negatively impacting hyper-resource-constrained usages (such as Raspberry Pis or whatnot) vs. grammars? That's what I'm unable to weigh -- I feel like that's a strategic decision that's a bit above my level.

In short, we don't need the cache for situations that don't use grammars, and we're adding a bit of memory usage (n_vocab*2) to every context that we're creating. On most systems this isn't a problem, but on highly-constrained systems (such as Raspberry Pi and whatnot) then this is wasted memory.

How do we weigh the interests of memory-constrained users vs. grammar-enabled users? That's something that I'm not able to make, but overall I think that improving speed performance on grammar-enabled sampling is going to benefit the largest number of people, and the ultra-constrained users are going to be pretty small. We might want to make a note somewhere in a comment that if one is looking for a way to decrease memory usage that they could disable the caching, but beyond that we're probably fine with kicking that can down the road.

Reply:
I added a log for the memory usage of the "token to piece" caches:

```bash
# llama 3
llm_load_vocab: token to piece cache size = 1.5928 MB
```

I think this is completely fine and no need to worry about it for now

Reply:
Excellent, thank you! That was the one reservation that held me back from fully approving #6811 (I felt that choice required someone with a larger project scope than I have), so I'm very happy to have you weigh in on that.